tests/test_typechecker.py

```diff
@@ -4668,6 +4668,71 @@ class Foo(typing.Generic[T]):
     def test_frozenset(self):
         self.assertTrue(pytypes.is_of_type(frozenset({1, 2, 'a', None, 'b'}), typing.AbstractSet[typing.Union[str, int, None]]))
 
+    # See: https://github.com/Stewori/pytypes/issues/32
+    # See: https://github.com/Stewori/pytypes/issues/33
+    def test_empty_values(self):
+        self.assertTrue(pytypes.is_of_type([], typing.Sequence))
+        self.assertTrue(pytypes.is_of_type([], typing.Sequence[int]))
+
+        for interface in (typing.Iterable, typing.Sized, typing.Container):
+            self.assertTrue(isinstance(set(), interface), interface)
+            self.assertTrue(pytypes.is_of_type(set(), interface), interface)
+            self.assertTrue(isinstance([], interface), interface)
+            self.assertTrue(pytypes.is_of_type([], interface), interface)
+
+    # See: https://github.com/Stewori/pytypes/issues/21
+    def test_tuple_ellipsis(self):
+        class Foo:
+            pass
+
+        self.assertTrue(pytypes.is_subtype(typing.Tuple[Foo], typing.Tuple[object, ...]))
+        self.assertTrue(pytypes.is_subtype(typing.Tuple[Foo], typing.Tuple[typing.Any, ...]))
+
+    # See: https://github.com/Stewori/pytypes/issues/24
+    def test_bound_typevars_readonly(self):
+        T = typing.TypeVar('T', covariant=True)
+
+        class L(typing.List[T]):
+            pass
+
+        C = typing.TypeVar('T', bound=L)
+
+        self.assertTrue(pytypes.is_subtype(L[float], C))
+        self.assertTrue(pytypes.is_subtype(L[float], C, bound_typevars={}))
+        self.assertFalse(pytypes.is_subtype(L[float], C, bound_typevars_readonly=True, bound_typevars={}))
+        self.assertTrue(pytypes.is_subtype(L[float], C, bound_typevars_readonly=False, bound_typevars={}))
+
+    # See: https://github.com/Stewori/pytypes/issues/22
+    def test_forward_declaration(self):
+        Wrapper = typing.Union[
+            typing.Sequence['Data'],
+        ]
+
+        Data = typing.Union[
+            Wrapper,
+            str, bytes, bool, float, int, dict,
+        ]
+
+        with self.assertRaises(pytypes.ForwardRefError):
+            pytypes.is_subtype(typing.Sequence[float], Wrapper)
+
+        pytypes.resolve_fw_decl(Wrapper)
+
+        self.assertTrue(pytypes.is_subtype(typing.Sequence[float], Wrapper))
+        self.assertTrue(pytypes.is_subtype(int, Data))
+        self.assertTrue(pytypes.is_subtype(float, Data))
+        self.assertFalse(pytypes.is_subtype(Data, Wrapper))
+        self.assertFalse(pytypes.is_subtype(Wrapper, Data))
```

Comment:
``self.assertFalse(pytypes.is_subtype(Wrapper, Data))`` is failing now. Unlike the line above, this is actually true. As participant of the Data-Union, Wrapper is indeed a subtype of it. As discussed earlier the converse is not true.

Reply:
Fixed. :-) I think I have no clue about typing anymore. ;-)

Reply:
Sorry to hear that. Hope it doesn't stop you from helping to improve pytypes ;)
Feel free to ask more detals if it helps. Here I think it boils down to union semantics, so let me point out the main rules again:
``Union[x, y]`` is subtype of ``z`` iff ``x`` is subtype of ``z`` *and* ``y`` is subtype of ``z``.
``z`` is subtype of ``Union[x, y]`` iff ``z`` is subtype of ``x`` *or* ``z`` is subtype of ``y``.
This directly corresponds to union operation from set theory if x, y, z are sets and subtype means subset. There is also the intersection correspondence in type theory, but PEP 484 does not provide an intersection type (was once discussed in an issue of typing, but then drpped because rare use cases did not outweight the difficulties in implementing it for static type checking/mypy)